File: app/javascript/controllers/journal_prompts_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["prompt", "contextBar"]

  static values = {
    todayTrades: { type: Number, default: 0 },
    todayPnl: { type: Number, default: 0 },
    todayWins: { type: Number, default: 0 },
    todayLosses: { type: Number, default: 0 },
    todaySymbols: { type: Array, default: [] },
    winRate: { type: Number, default: 0 },
    currentStreak: { type: Number, default: 0 },
    streakType: { type: String, default: "" },
    bestSymbol: { type: String, default: "" },
    bestPnl: { type: Number, default: 0 },
    worstSymbol: { type: String, default: "" },
    worstPnl: { type: Number, default: 0 }
  }

  static basePrompts = {
    content: [
      "What was the most significant trade you made today and why?",
      "Describe your emotional state during the trading session.",
      "What surprised you about the market today?",
      "Which of your setups worked best today? Why?",
      "Did you deviate from your plan? What triggered the deviation?",
      "Rate your discipline today from 1-10. What influenced the score?",
      "What pattern did you notice in price action today?",
      "How did pre-market preparation help (or hurt) your trading?"
    ],
    market_conditions: [
      "Was the market trending, ranging, or choppy today?",
      "Which sectors led or lagged today?",
      "Were there any significant news events affecting the market?",
      "How was overall volume compared to the average?",
      "Describe the VIX / volatility environment.",
      "Were key support/resistance levels respected today?"
    ],
    plan: [
      "What setups are you watching for tomorrow?",
      "What is your maximum risk for the day?",
      "Which symbols are on your focus list?",
      "What are the key levels you need to watch?",
      "Are there any economic events to be aware of?",
      "What is your primary strategy for tomorrow's session?"
    ],
    review: [
      "Did you follow your trading plan? Grade yourself A-F.",
      "What is one thing you would do differently?",
      "Were your position sizes appropriate for the setups?",
      "Did you exit trades at the right time? Too early or too late?",
      "What was your best decision today? Worst decision?",
      "What lesson will you carry into tomorrow?"
    ]
  }

  connect() {
    this.showPrompts()
    this.renderContextBar()
  }

  getSmartPrompts() {
    const prompts = {
      content: [...this.constructor.basePrompts.content],
      market_conditions: [...this.constructor.basePrompts.market_conditions],
      plan: [...this.constructor.basePrompts.plan],
      review: [...this.constructor.basePrompts.review]
    }

    const trades = this.todayTradesValue
    const pnl = this.todayPnlValue
    const wins = this.todayWinsValue
    const losses = this.todayLossesValue
    const symbols = this.todaySymbolsValue
    const streak = this.currentStreakValue
    const streakType = this.streakTypeValue

    if (trades === 0) {
      // No trades today
      prompts.content.unshift(
        "You haven't traded today. What kept you on the sidelines?",
        "Was not trading today a deliberate decision or did you miss opportunities?"
      )
      prompts.plan.unshift(
        "Since you didn't trade today, what are you looking for tomorrow?",
        "Use this down time to review your playbook. Which setups need refinement?"
      )
    } else if (pnl > 0) {
      // Green day
      prompts.content.unshift(
        `You're up $${pnl.toFixed(2)} today across ${trades} trade${trades > 1 ? 's' : ''}. What went right?`,
        `${wins} winning trade${wins > 1 ? 's' : ''} today. Was it skill, setup quality, or market conditions?`
      )
      if (this.bestSymbolValue) {
        prompts.content.unshift(
          `${this.bestSymbolValue} was your best trade (+$${this.bestPnlValue.toFixed(2)}). Walk through your decision process.`
        )
      }
      prompts.review.unshift(
        "Great day! But were you over-leveraged, or was your sizing appropriate?",
        "Would you take the same trades tomorrow? What made today's setups compelling?"
      )
    } else if (pnl < 0) {
      // Red day
      prompts.content.unshift(
        `Down $${Math.abs(pnl).toFixed(2)} today. What would you change if you could replay the session?`,
        `${losses} losing trade${losses > 1 ? 's' : ''} today. Were these valid setups that didn't work, or forced trades?`
      )
      if (this.worstSymbolValue) {
        prompts.content.unshift(
          `${this.worstSymbolValue} was your worst trade ($${this.worstPnlValue.toFixed(2)}). Did you follow your stop?`
        )
      }
      prompts.review.unshift(
        "Red days happen. What's the one adjustment that would have improved today?",
        "Did you cut losses quickly or let them run? Be honest with yourself."
      )
      prompts.plan.unshift(
        "After a losing day, what will you do differently tomorrow?",
        "Should you reduce size tomorrow or trade your normal plan?"
      )
    }

    // Streak-specific prompts
    if (streak >= 3 && streakType === "win") {
      prompts.content.unshift(
        `You're on a ${streak}-trade win streak. Are you staying disciplined or getting overconfident?`
      )
      prompts.plan.unshift(
        "Hot streaks can breed complacency. What's your plan to stay sharp?"
      )
    } else if (streak >= 3 && streakType === "loss") {
      prompts.content.unshift(
        `You're in a ${streak}-trade losing streak. Is this a setup issue, sizing issue, or market issue?`
      )
      prompts.plan.unshift(
        "Consider reducing size or taking a break. What's your recovery plan?",
        "After a losing streak, are you revenge trading or staying process-focused?"
      )
    }

    // Symbol-specific prompts
    if (symbols.length > 0) {
      const symList = symbols.slice(0, 3).join(", ")
      prompts.market_conditions.unshift(
        `How did ${symList} behave relative to the broader market today?`
      )
    }

    if (symbols.length > 4) {
      prompts.review.unshift(
        `You traded ${symbols.length} different symbols. Was this overtrading or valid diversification?`
      )
    }

    return prompts
  }

  showPrompts() {
    const prompts = this.getSmartPrompts()
    this.promptTargets.forEach(el => {
      const field = el.dataset.field
      const fieldPrompts = prompts[field]
      if (fieldPrompts && fieldPrompts.length > 0) {
        // Show the first (most contextual) prompt
        el.textContent = fieldPrompts[0]
        el.style.display = ""
      }
    })
  }

  refresh() {
    const prompts = this.getSmartPrompts()
    this.promptTargets.forEach(el => {
      const field = el.dataset.field
      const fieldPrompts = prompts[field]
      if (fieldPrompts && fieldPrompts.length > 0) {
        const randomIndex = Math.floor(Math.random() * fieldPrompts.length)
        el.textContent = fieldPrompts[randomIndex]
      }
    })
  }

  renderContextBar() {
    if (!this.hasContextBarTarget) return
    const trades = this.todayTradesValue
    if (trades === 0) {
      this.contextBarTarget.innerHTML = `
        <div style="display: flex; align-items: center; gap: 0.5rem; padding: 0.75rem 1rem; background: var(--bg); border-radius: var(--radius); margin-bottom: 1rem; font-size: 0.8125rem;">
          <span class="material-icons-outlined" style="font-size: 1.125rem; color: var(--text-secondary);">info</span>
          <span>No trades recorded today. Prompts are tailored to non-trading days.</span>
        </div>
      `
      return
    }

    const pnl = this.todayPnlValue
    const wins = this.todayWinsValue
    const losses = this.todayLossesValue
    const pnlColor = pnl >= 0 ? 'var(--positive)' : 'var(--negative)'
    const pnlSign = pnl >= 0 ? '+' : ''

    this.contextBarTarget.innerHTML = `
      <div style="display: flex; align-items: center; gap: 1.5rem; padding: 0.75rem 1rem; background: var(--bg); border-radius: var(--radius); margin-bottom: 1rem; font-size: 0.8125rem; flex-wrap: wrap;">
        <span style="display: flex; align-items: center; gap: 0.375rem;">
          <span class="material-icons-outlined" style="font-size: 1rem; color: ${pnlColor};">${pnl >= 0 ? 'trending_up' : 'trending_down'}</span>
          <strong>Today:</strong>
          <span style="color: ${pnlColor}; font-weight: 600;">${pnlSign}$${Math.abs(pnl).toFixed(2)}</span>
        </span>
        <span>${trades} trade${trades > 1 ? 's' : ''}</span>
        <span style="color: var(--positive);">${wins}W</span>
        <span style="color: var(--negative);">${losses}L</span>
        <span style="color: var(--text-secondary); font-size: 0.75rem;">Prompts tailored to your session</span>
      </div>
    `
  }
}

File: app/controllers/journal_prompts_controller.rb
class JournalPromptsController < ApplicationController
  include ApiConnected

  def show
    stats_thread = Thread.new do
      api_client.overview
    rescue => e
      Rails.logger.error("journal_prompts stats: #{e.message}")
      {}
    end

    streaks_thread = Thread.new do
      api_client.streaks
    rescue => e
      Rails.logger.error("journal_prompts streaks: #{e.message}")
      {}
    end

    trades_thread = Thread.new do
      api_client.trades(per_page: 10, sort: "closed_at", direction: "desc")
    rescue => e
      Rails.logger.error("journal_prompts trades: #{e.message}")
      {}
    end

    stats = stats_thread.value || {}
    streaks = streaks_thread.value || {}
    raw_trades = trades_thread.value || {}

    trades = raw_trades.is_a?(Hash) ? (raw_trades["trades"] || raw_trades["data"] || []) : Array(raw_trades)
    daily_pnl = normalize_daily_pnl(stats)

    cs = streaks.is_a?(Hash) ? streaks["current_streak"] : nil
    streak_count = cs.is_a?(Hash) ? cs["count"].to_i : cs.to_i
    streak_type = cs.is_a?(Hash) ? cs["type"] : (streaks.is_a?(Hash) ? streaks["streak_type"] : nil)

    today_pnl = daily_pnl[Date.today.to_s].to_f
    yesterday_pnl = daily_pnl[(Date.today - 1).to_s].to_f
    today_trades = trades.select { |t| t["closed_at"]&.start_with?(Date.today.to_s) }

    @context = {
      today_pnl: today_pnl,
      yesterday_pnl: yesterday_pnl,
      today_trade_count: today_trades.length,
      streak_count: streak_count,
      streak_type: streak_type,
      win_rate: stats["win_rate"]&.to_f,
      total_trades: stats["total_trades"].to_i,
      recent_symbols: trades.first(5).map { |t| t["symbol"] }.compact.uniq
    }

    @prompts = generate_prompts(@context)
    @categories = @prompts.map { |p| p[:category] }.uniq
  end

  private

  def normalize_daily_pnl(stats)
    raw = stats.is_a?(Hash) ? (stats["daily_pnl"] || {}) : {}
    pnl = raw.is_a?(Array) ? raw.to_h : raw
    pnl.transform_keys(&:to_s).transform_values(&:to_f)
  end

  def generate_prompts(ctx)
    prompts = []

    # Always-available prompts
    prompts.concat(core_prompts)

    # Context-driven prompts
    if ctx[:today_pnl] > 0
      prompts.concat(winning_day_prompts(ctx))
    elsif ctx[:today_pnl] < 0
      prompts.concat(losing_day_prompts(ctx))
    else
      prompts.concat(no_trades_prompts)
    end

    if ctx[:streak_count].abs >= 3
      prompts.concat(streak_prompts(ctx))
    end

    if ctx[:win_rate] && ctx[:win_rate] < 45
      prompts.concat(struggling_prompts)
    end

    prompts.concat(reflection_prompts)
    prompts.concat(planning_prompts)
    prompts.concat(psychology_prompts)

    prompts.uniq { |p| p[:text] }
  end

  def core_prompts
    [
      { category: "Daily Review", icon: "today", color: "#1a73e8",
        text: "What was my best decision today and why?",
        follow_up: "How can I replicate this decision-making process?" },
      { category: "Daily Review", icon: "today", color: "#1a73e8",
        text: "What was my worst decision today? What triggered it?",
        follow_up: "What rule would have prevented this?" },
      { category: "Daily Review", icon: "today", color: "#1a73e8",
        text: "Did I follow my trading plan today? Rate yourself 1-10.",
        follow_up: "What specific steps will improve this score tomorrow?" },
      { category: "Daily Review", icon: "today", color: "#1a73e8",
        text: "What emotional state was I in during my trades?",
        follow_up: "How did my emotions impact my entries and exits?" }
    ]
  end

  def winning_day_prompts(ctx)
    [
      { category: "Winning Day", icon: "celebration", color: "var(--positive)",
        text: "I made #{ActionController::Base.helpers.number_to_currency(ctx[:today_pnl])} today. Was this skill or luck?",
        follow_up: "What percentage of today's gains came from following my system?" },
      { category: "Winning Day", icon: "celebration", color: "var(--positive)",
        text: "Am I at risk of overconfidence after today's win?",
        follow_up: "What's my plan to stay disciplined tomorrow?" },
      { category: "Winning Day", icon: "celebration", color: "var(--positive)",
        text: "Which of today's trades would I take again? Which would I skip?",
        follow_up: "What separates the A+ setups from the rest?" }
    ]
  end

  def losing_day_prompts(ctx)
    [
      { category: "Losing Day", icon: "healing", color: "var(--negative)",
        text: "I lost #{ActionController::Base.helpers.number_to_currency(ctx[:today_pnl].abs)} today. Was my process still good?",
        follow_up: "Good process with bad outcomes is still good trading. Separate the two." },
      { category: "Losing Day", icon: "healing", color: "var(--negative)",
        text: "Did I honor my stop losses today or did I let losers run?",
        follow_up: "Write down the exact moment you should have exited vs when you did." },
      { category: "Losing Day", icon: "healing", color: "var(--negative)",
        text: "Was I revenge trading at any point today?",
        follow_up: "What was the trigger? How will I interrupt this pattern next time?" },
      { category: "Losing Day", icon: "healing", color: "var(--negative)",
        text: "What would I tell a friend who had today's results?",
        follow_up: "Practice the same self-compassion you'd offer them." }
    ]
  end

  def no_trades_prompts
    [
      { category: "No Trades", icon: "hourglass_empty", color: "#9e9e9e",
        text: "I didn't trade today. Was this patience or avoidance?",
        follow_up: "Sometimes the best trade is no trade. Was this intentional?" },
      { category: "No Trades", icon: "hourglass_empty", color: "#9e9e9e",
        text: "What setups did I see but pass on? Why?",
        follow_up: "Review these later — was passing the right call?" }
    ]
  end

  def streak_prompts(ctx)
    type = ctx[:streak_type]&.downcase == "loss" ? "losing" : "winning"
    [
      { category: "Streak", icon: "local_fire_department", color: "#e65100",
        text: "I'm on a #{ctx[:streak_count].abs}-trade #{type} streak. How is this affecting my mindset?",
        follow_up: "Streaks are normal. What matters is whether your process remains consistent." },
      { category: "Streak", icon: "local_fire_department", color: "#e65100",
        text: "Should I adjust my position sizing during this #{type} streak?",
        follow_up: "Consider reducing size after losses to protect capital, or taking some profits during wins." }
    ]
  end

  def struggling_prompts
    [
      { category: "Performance", icon: "trending_down", color: "#f9a825",
        text: "My win rate is below 45%. Is my strategy still viable?",
        follow_up: "Review: Is the problem entries, exits, or trade selection?" },
      { category: "Performance", icon: "trending_down", color: "#f9a825",
        text: "Should I paper trade for a week to rebuild confidence?",
        follow_up: "There's no shame in stepping back to recalibrate." }
    ]
  end

  def reflection_prompts
    [
      { category: "Reflection", icon: "psychology", color: "#9c27b0",
        text: "What belief about the market was challenged today?",
        follow_up: "How does this change your approach going forward?" },
      { category: "Reflection", icon: "psychology", color: "#9c27b0",
        text: "If I could go back to this morning, what one thing would I change?",
        follow_up: "This is your most actionable improvement for tomorrow." },
      { category: "Reflection", icon: "psychology", color: "#9c27b0",
        text: "What am I grateful for in my trading journey right now?",
        follow_up: "Gratitude builds resilience during drawdowns." },
      { category: "Reflection", icon: "psychology", color: "#9c27b0",
        text: "Describe your ideal trading day. How close was today?",
        follow_up: "What's the gap between ideal and reality? What's one step to close it?" }
    ]
  end

  def planning_prompts
    [
      { category: "Planning", icon: "event_note", color: "#00897b",
        text: "What are my top 3 watchlist symbols for tomorrow?",
        follow_up: "Why these? What's the thesis for each?" },
      { category: "Planning", icon: "event_note", color: "#00897b",
        text: "What is my maximum acceptable loss tomorrow?",
        follow_up: "Write it down. Commit to it. Set an alarm." },
      { category: "Planning", icon: "event_note", color: "#00897b",
        text: "What economic events or catalysts should I be aware of tomorrow?",
        follow_up: "Check the economic calendar and earnings schedule." }
    ]
  end

  def psychology_prompts
    [
      { category: "Psychology", icon: "self_improvement", color: "#5d4037",
        text: "On a scale of 1-10, how disciplined was I today?",
        follow_up: "What specific action would bump this score up by 1?" },
      { category: "Psychology", icon: "self_improvement", color: "#5d4037",
        text: "Did I experience FOMO today? How did I handle it?",
        follow_up: "Write down the specific moment and how you responded." },
      { category: "Psychology", icon: "self_improvement", color: "#5d4037",
        text: "What physical state was I in today? (Sleep, exercise, nutrition)",
        follow_up: "How does your physical well-being correlate with trading performance?" }
    ]
  end
end
